src/utils/server-utils.ts

```diff
@@ -0,0 +1,165 @@
+import { xml2js } from 'xml-js';
+import { Map } from 'immutable';
+import { merge, unset, get, isString, union, isEmpty } from 'lodash';
+import { format, parse } from 'url';
+
+import config from '../config/prism.json';
+import { AvailableDates } from '../config/types';
+
+const xml2jsOptions = {
+  compact: true,
+  trim: true,
+  ignoreComment: true,
+};
+
+export function formatServerUri(
+  serverUri: string,
+  queryProp: { [key: string]: string | boolean | number },
+  selectedDate?: number,
+) {
+  // The second arg of 'parse' allows us to have 'query' as an object
+  const { query, ...parsedUrl } = parse(serverUri, true);
+
+  // Removing 'search' to be able to format by 'query'
+  unset(parsedUrl, 'search');
+
+  const selectedDateString = selectedDate
+    ? moment(selectedDate).format('YYYY-MM-DD')
+    : '';
+
+  return decodeURI(
+    format({
+      ...parsedUrl,
+      query: merge(query, queryProp, { time: selectedDateString }),
+    }),
+  );
+}
+
+/**
+ * Format the raw data to { [layerId]: availableDates }
+ * @param rawLayers Layers data return by the server 'GetCapabilities' request
+ * @param layerIdPath path to layer's id
+ * @param datesPath path to layer's available dates
+ * @returns an object shape like { [layerId]: availableDates }
+ */
+function formatCapabilitiesInfo(
+  rawLayers: any,
+  layerIdPath: string,
+  datesPath: string,
+): { [key: string]: number[] } {
+  return rawLayers.reduce((acc: any, layer: any) => {
+    const layerId = get(layer, layerIdPath);
+    const rawDates = get(layer, datesPath, []);
+
+    const dates: (string | { _text: string })[] = isString(rawDates)
+      ? rawDates.split(',')
+      : rawDates;
+
+    const availableDates = dates
+      .filter(date => !isEmpty(date))
+      .map(date => moment(get(date, '_text', date)).valueOf());
+
+    if (Object.prototype.hasOwnProperty.call(acc, layerId)) {
+      const { [layerId]: oldLayerDates } = acc;
+      return {
+        ...acc,
+        [layerId]: union(availableDates, oldLayerDates),
+      };
+    }
+
+    return {
+      ...acc,
+      [layerId]: availableDates,
+    };
+  }, {});
+}
+
+/**
+ * List capabilities for a WMS layer.
+ * @param serverUri
+ */
+async function getWMSCapabilities(serverUri: string) {
+  const requestUri = formatServerUri(serverUri, { request: 'GetCapabilities' });
+
+  return new Promise(resolve => {
+    fetch(requestUri)
+      .then(response => response.text())
+      .then(responseText => xml2js(responseText, xml2jsOptions))
+      .then(responseJs => {
+        const rawLayers = get(responseJs, 'WMS_Capabilities.Capability.Layer');
+
+        const flattenLayers = Array.isArray(rawLayers)
+          ? rawLayers.reduce((acc, { Layer }) => acc.concat(Layer), [])
+          : get(rawLayers, 'Layer', []);
+
+        const layers = formatCapabilitiesInfo(
+          flattenLayers,
+          'Name._text',
+          'Dimension._text',
+        );
+
+        resolve(Map(layers));
+      })
```

Comment:
We're _generally_ using `async-await`, any particular reason you're using a Promise chain here?

Reply:
Not a big reason ... I prefer the `then/catch` syntax than `try/catch` (I think it more readable and less `await`)

But less make that more conventional.